Reading *guns.csv* file and creating a list of lists with each row in the file a sub-list in the outer list

In [94]:
import csv
data = list(csv.reader(open("guns.csv", "r")))
print (data[0:10])

[['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education'], ['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'], ['5', '2012', '02', 'Suicide', '0', 'M', '31', 'White', '100', 'Other specified', '2'], ['6', '2012', '02', 'Suicide', '0', 'M', '17', 'Native American/Native Alaskan', '100', 'Home', '1'], ['7', '2012', '02', 'Undetermined', '0', 'M', '48', 'White', '100', 'Home', '2'], ['8', '2012', '03', 'Suicide', '0', 'M', '41', 'Native American/Native Alaskan', '100', 'Home', '2'], ['9', '2012', '02', 'Accidental', '0', 'M', '50', 'White', '100', 'Other specified', '3']]


Separating the header row from the records 

In [95]:
header = data[0]
data = data[1:]
print (header)

['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education']


Retrieving the years for which the data is provided:

In [96]:
years = {row[1] for row in data}
print (years)

set(['2014', '2013', '2012'])


The data has been provided for the years **2012**, **2013**, and **2014**. 

Retrieving the intents behind the gun deaths listed:

In [97]:
intents = {row[3] for row in data}
print (intents)

set(['Undetermined', 'Homicide', 'Suicide', 'Accidental', 'NA'])


The intents behind the gun deaths recorded are **Undetermined**, **Homicide**, **Suicide**, **Accidental**, and **NA**.  

Retrieving the races included in the data:

In [98]:
races = {row[7] for row in data}
print (races)

set(['Native American/Native Alaskan', 'Black', 'White', 'Asian/Pacific Islander', 'Hispanic'])


The races recorded in the data: **Native American/Native Alaskan**, **Black**, **White**, **Asian/Pacific Islander**, and **Hispanic**. 

Retrieving the various places where the gun death incidents were reported:

In [99]:
places = {row[9] for row in data}
print (places)

set(['Farm', 'NA', 'Other unspecified', 'Industrial/construction', 'Trade/service area', 'Street', 'School/instiution', 'Home', 'Other specified', 'Sports', 'Residential institution'])


The gun death incidents have occurred **at Farms, on Industrial/Construction sites, in Trade and service areas, on streets, at schools and educational institutions, at homes, and at Residential institutions.** Some locations are **unspecified** and others mentioned as **NA**. **Sports** related gun deaths have also been recorded.   

Retrieving the educational status of the victims:

In [100]:
educational_status = {row[10] for row in data}
print (educational_status)

set(['NA', '1', '3', '2', '5', '4'])


The **educational status** of the victim coded as follows:   
1 - Less than High School    
2 - Graduated from High School or equivalent  
3 - Some College  
4 - At least graduated from college  
5 - Not Available   
NA - Not Applicable   

Finding the number of gun deaths in each of the given years:

In [101]:
year_counts = {}
for row in data:
    year = row[1]
    if year in year_counts:
        year_counts[year] += 1
    else:
        year_counts[year] = 1
print (year_counts)

{'2014': 33599, '2013': 33636, '2012': 33563}


Gun deaths recorded in the years **2012**, **2013**, and **2014** are **33563**, **33636**, and **33599**, respectively. We can infer that there hasn't been any significant change in the number of gun deaths from the year 2012 to 2014. 


Now we will find the number of male and female victims of gun deaths in each of the given three years: 

In [102]:
sexes = {row[5] for row in data}
print (sexes)

set(['M', 'F'])


In [103]:
def year_sex_count(year):
    count = {'M':0, 'F':0}
    for row in data:
        if row[1] == year:
            gender = row[5]
            count[gender] += 1
    return count

In [104]:
sex_count_2012 = year_sex_count('2012')
print (sex_count_2012)

sex_count_2013 = year_sex_count('2013')
print (sex_count_2013)

sex_count_2014 = year_sex_count('2014')
print (sex_count_2014)

{'M': 28838, 'F': 4725}
{'M': 28794, 'F': 4842}
{'M': 28717, 'F': 4882}


We can roughly say from the above numbers that the number of male casualties of gun deaths is way more than the number of female casualties. However, to come to such a conclusion, we will still need data on the female and the male population off the total population of the US.  
From the number of female casualties, we see that the number has slightly risen in 2013 to 4842 against the 4725 recorded the previous year.  

In [105]:
sex_counts = {'M': sex_count_2012['M'] + sex_count_2013['M'] + sex_count_2014['M'], 'F': sex_count_2012['F'] + sex_count_2013['F'] + sex_count_2014['F']}
print sex_counts

{'M': 86349, 'F': 14449}


The total number of male and female casualties of gun deaths between 2012 and 2014 are **86349** and **14449**, respectively. 

Finding the oldest victims of gun deaths:

In [106]:
#replacing 'NA' and '0' values for the age column with None
for row in data:
    if row[6] == 'NA' or row[6] == '0':
        row[6] = None

#Fetching the age of all the listed victims and storing it in a set
ages = {int(row[6]) if row[6] is not None else None for row in data}

#finding the age of the oldest victim
max_age = 0
for age in ages: 
    if age is not None and age > max_age:
        max_age = age
        
#fetching the details of the oldest victims of gun deaths     
oldest_victims = filter(lambda row: row[6] is not None and int(row[6])==max_age, data)
print (oldest_victims)

[['34903', '2013', '09', 'Homicide', '0', 'M', '107', 'Black', '100', 'Home', '1']]


The oldest victim of gun deaths is reported to be a **Black Male**, aged **107**. The intent behind the gun violence is **Homicide**. The victim was killed in his home in September 2013.  

Finding the youngest victims of gun deaths: 

In [107]:
#finding the age of the youngest victims
min_age = iter(ages).next()
for age in ages:
    if age is not None and age < min_age:
        min_age = age

#fetching the details of the youngest victims of gun deaths     
youngest_victims = filter(lambda row: row[6] is not None and int(row[6])==min_age, data)
print (youngest_victims[0:5])

[['1770', '2012', '05', 'Homicide', '0', 'F', '1', 'Hispanic', '270', 'Home', '5'], ['2057', '2012', '08', 'Homicide', '0', 'M', '1', 'Hispanic', '210', 'Home', '1'], ['2492', '2012', '01', 'Homicide', '0', 'F', '1', 'Hispanic', '210', 'Home', '1'], ['3651', '2012', '06', 'Homicide', '0', 'M', '1', 'Hispanic', '210', 'Street', '1'], ['13224', '2012', '04', 'Accidental', '0', 'F', '1', 'Hispanic', '210', 'Home', '1']]


In [108]:
#finding the intent behind the gun deaths of the youngest victims
intent_youngest_victims = {row[3] for row in youngest_victims}
print intent_youngest_victims

set(['Homicide', 'Accidental'])


In [109]:
#the number of youngest victims of gun deaths
num_youngest_victims = len(youngest_victims)
print num_youngest_victims

38


In [110]:
#number of homicide and accidental cases off the total number of cases involving the youngest victims
count_intent_youngest_victims = {intent:0 for intent in intent_youngest_victims}

for row in youngest_victims:
    intent = row[3]
    count_intent_youngest_victims[intent] += 1

print count_intent_youngest_victims    
        

{'Homicide': 34, 'Accidental': 4}


In [111]:
#finding the number of homicide cases, off the 34, by race of the youngest casualities
race_homicide_count_youngest_victims = {}
for row in youngest_victims:
    intent = row[3]
    race = row[7]
    if intent == 'Homicide':
        if race not in race_homicide_count_youngest_victims:
            race_homicide_count_youngest_victims[race] = 1
        else:
            race_homicide_count_youngest_victims[race] += 1
print race_homicide_count_youngest_victims

{'Hispanic': 7, 'White': 9, 'Asian/Pacific Islander': 2, 'Black': 16}


A total of **38** cases have been recorded where the victims are **aged 1**; the youngest casualties of gun deaths. Off the 38 cases, **34** are **homicides** and **4** are **accidental deaths**. And off the 34 homicide victims, **16** were of the *Black* descent, 9 White, 7 Hispanic, and 2 Asian/Pacific Islander.   

Now we will find the number of cases which involved a police officer in the shooting:

In [112]:
#filtering out the rows with the fifth column having '1' for value
police_shooting = [row for row in data if row[4] == '1']
print police_shooting[0:5]

[['62', '2012', '08', 'Homicide', '1', 'M', '28', 'White', '100', 'NA', '2'], ['84', '2012', '09', 'Homicide', '1', 'M', '34', 'White', '100', 'NA', '2'], ['93', '2012', '10', 'Homicide', '1', 'M', '24', 'Native American/Native Alaskan', '100', 'NA', '1'], ['635', '2012', '06', 'Homicide', '1', 'M', '22', 'Black', '100', 'NA', '2'], ['1005', '2012', '01', 'Homicide', '1', 'M', '36', 'White', '100', 'NA', '3']]


In [113]:
total_cases = len(data)
print total_cases

police_count = len(police_shooting)
print police_count

100798
1402


In [114]:
#finding the count of gun deaths involving a police officer against each race
race_police_count = {}
for row in police_shooting:
    race = row[7]
    if race not in race_police_count:
        race_police_count[race] = 1
    else:
        race_police_count[race] += 1
print race_police_count

{'Native American/Native Alaskan': 25, 'Hispanic': 282, 'White': 709, 'Asian/Pacific Islander': 30, 'Black': 356}


Off the **100798** gun deaths recorded over the three years, **1402** cases involved a police in the shooting. **709** such cases involved victims of white descent, **356** of Black descent, and **282** of Hispanic descent.  


Now we will find the number of gun deaths by place of the incident:

In [115]:
place_count = {}
for row in data:
    place = row[9]
    if place not in place_count:
        place_count[place] = 1
    else:
        place_count[place] += 1
print place_count

{'Farm': 470, 'NA': 1384, 'Other unspecified': 8867, 'Industrial/construction': 248, 'Trade/service area': 3439, 'Street': 11151, 'School/instiution': 671, 'Home': 60486, 'Other specified': 13751, 'Sports': 128, 'Residential institution': 203}


The numbers suggest that by far most incidents of gun deaths were reported to have occurred at home. 

In [116]:
#funtion to count by place and intent
def place_intent_count(place):
    count = {}
    for row in data: 
        if row[9] == place:
            intent = row[3]
            if intent not in count:
                count[intent] = 1
            else:
                count[intent] += 1
    return count

In [117]:
#the count by intent of the gun deaths at home
home_intent_count = place_intent_count('Home')
print home_intent_count

{'Undetermined': 535, 'Homicide': 13613, 'Suicide': 45415, 'Accidental': 923}


And most deaths that occurred at **home** were a result of the victims taking their own lives. The number of suicides recorded is **45415**, followed by Homicide at **13613**.   

In [118]:
#the count by intent of gun deaths at schools and educational institutions
school_intent_count = place_intent_count('School/instiution')
print school_intent_count

{'Homicide': 361, 'Undetermined': 4, 'Suicide': 295, 'Accidental': 11}


The gun deaths that occurred at **schools** and **educational institutions** were largely due to **Homicide (361 cases)** and **Suicide (295 cases)**. 

In [119]:
#funtion to count by the race of the victim, the intent, and the place
def place_intent_race_count(place, intent):
    count={}
    for row in data:
        if row[9] == place and row[3] == intent:
            race = row[7]
            if race not in count:
                count[race] = 1
            else:
                count[race] += 1
    return count

In [120]:
home_suicide_race_count = place_intent_race_count('Home', 'Suicide')
print home_suicide_race_count

{'Native American/Native Alaskan': 422, 'Black': 2270, 'White': 39993, 'Asian/Pacific Islander': 483, 'Hispanic': 2247}


Off the **45415** cases of **suicides** reported at **Home**, **39993**, i.e. **88%**, of the victims were **Whites**.  

In [121]:
home_homicide_race_count = place_intent_race_count('Home', 'Homicide')
print home_homicide_race_count

{'Native American/Native Alaskan': 147, 'Black': 6413, 'White': 4905, 'Asian/Pacific Islander': 199, 'Hispanic': 1949}


Off the **13613** cases of **Homicides** at **Home**, **6413** of the victims were **Blacks**, **4905 Whites**, and **1949 Hispanics**. 

In [122]:
school_homicide_race_count = place_intent_race_count('School/instiution', 'Homicide')
print school_homicide_race_count

{'Native American/Native Alaskan': 3, 'Asian/Pacific Islander': 18, 'White': 104, 'Black': 201, 'Hispanic': 35}


Off the **361** cases of **homicides**, the foremost cause of gun deaths, at **schools** and **educational institutions**, **201** of the victims were Blacks. 

In [123]:
school_suicide_race_count = place_intent_race_count('School/instiution', 'Suicide')
print school_suicide_race_count

{'Native American/Native Alaskan': 2, 'Hispanic': 15, 'White': 239, 'Asian/Pacific Islander': 11, 'Black': 28}


Off the **295** cases of **Suicides** reported at **schools** and **educational institutions**, **239**, i.e. **81%**, of the victims were **Whites**. 

Now we will find the count of gun deaths by race:

In [124]:
race_counts = {}
for row in data:
    race = row[7]
    if race in race_counts:
        race_counts[race] += 1
    else:
        race_counts[race] = 1

print race_counts

{'Native American/Native Alaskan': 917, 'Black': 23296, 'White': 66237, 'Asian/Pacific Islander': 1326, 'Hispanic': 9022}


Just by the above numbers, we cannot come to any conclusion as to which race has seen the most gun deaths. 
We will read another file *census.csv* that has data pertaining to the population of different races off the total population of the US. From that data, we will compute the number of gun deaths per 100000 people in each race.  

In [125]:
census = list(csv.reader(open("census.csv")))
print census
print '\n'

#separating the header row and assigning it to a list
census_header = census[0]
print census_header
print '\n'

census = census[1]
print census

[['Id', 'Year', 'Id', 'Sex', 'Id', 'Hispanic Origin', 'Id', 'Id2', 'Geography', 'Total', 'Race Alone - White', 'Race Alone - Hispanic', 'Race Alone - Black or African American', 'Race Alone - American Indian and Alaska Native', 'Race Alone - Asian', 'Race Alone - Native Hawaiian and Other Pacific Islander', 'Two or More Races'], ['cen42010', 'April 1, 2010 Census', 'totsex', 'Both Sexes', 'tothisp', 'Total', '0100000US', '', 'United States', '308745538', '197318956', '44618105', '40250635', '3739506', '15159516', '674625', '6984195']]


['Id', 'Year', 'Id', 'Sex', 'Id', 'Hispanic Origin', 'Id', 'Id2', 'Geography', 'Total', 'Race Alone - White', 'Race Alone - Hispanic', 'Race Alone - Black or African American', 'Race Alone - American Indian and Alaska Native', 'Race Alone - Asian', 'Race Alone - Native Hawaiian and Other Pacific Islander', 'Two or More Races']


['cen42010', 'April 1, 2010 Census', 'totsex', 'Both Sexes', 'tothisp', 'Total', '0100000US', '', 'United States', '308745538'

We will manually create a mapping of the race and the population count of that race from the census. Each key in the race_counts will be the key in the dictionary, as the race names in the census and the gun deaths data differ.  

In [126]:
mapping = {'Asian/Pacific Islander': int(census[15])+int(census[14]), 'White':int(census[10]), 'Hispanic':int(census[11]), 'Black':int(census[12]), 'Native American/Native Alaskan': int(census[13])} 
print mapping

{'Native American/Native Alaskan': 3739506, 'Hispanic': 44618105, 'White': 197318956, 'Asian/Pacific Islander': 15834141, 'Black': 40250635}


In [127]:
race_count_per_hundredk = {race: float(count*100000)/mapping[race] for race, count in race_counts.items()}
print (race_count_per_hundredk)

{'Native American/Native Alaskan': 24.521955573811088, 'Asian/Pacific Islander': 8.374309664161762, 'White': 33.56849303419181, 'Black': 57.877347773519595, 'Hispanic': 20.220491210910907}


The above numbers depict the *rate of gun deaths per 100000 people of each racial category*. 

In [130]:
#homicide counts by race
homicide_race_counts = {}
for row in data:
    race = row[7]
    intent = row[3]
    if intent == 'Homicide':
        if race in homicide_race_counts:
            homicide_race_counts[race] += 1
        else:
            homicide_race_counts[race] = 1
print homicide_race_counts

{'Native American/Native Alaskan': 326, 'Black': 19510, 'White': 9147, 'Asian/Pacific Islander': 559, 'Hispanic': 5634}


In [131]:
homicide_race_per_hundredk = {race: float(count*100000)/mapping[race] for race,count in homicide_race_counts.items()}
print homicide_race_per_hundredk

{'Native American/Native Alaskan': 8.717729026240365, 'Asian/Pacific Islander': 3.5303462309701548, 'White': 4.6356417981453335, 'Black': 48.47128498718095, 'Hispanic': 12.627161104219912}


The above numbers depict the *rate of gun deaths due to homicide per 100000 people of each racial category*. Going by the numbers, *48.47* out of 100000 blacks were killed in gun violence between 2012 and 2014.  